pkg/virtctl/expose/expose.go

```diff
@@ -0,0 +1,211 @@
+	cmd.Flags().StringVar(&portName, "port-name", "", "Name of the port. Optional.")
+	cmd.SetUsageTemplate(templates.UsageTemplate())
+
+	return cmd
+}
+
+func usage() string {
+	usage := `  # Expose SSH to a virtual machine called 'myvm' as a node port (5555) of the cluster:
+  virtctl expose vm myvm --port=5555 --target-port=22 --name=myvm-ssh --type=NodePort")`
+	return usage
+}
+
+// executing the "expose" command
+func (o *Command) RunE(cmd *cobra.Command, args []string) error {
+	// first argument is type of VM: VM, offline VM or replica set VM
+	vmType := strings.ToLower(args[0])
+	// second argument must be name of the VM
+	vmName := args[1]
+
+	// these are used to convert the flag values into service spec values
+	var protocol v1.Protocol
+	var targetPort intstr.IntOrString
+	var serviceType v1.ServiceType
+
+	// convert from integer to the IntOrString type
+	targetPort = intstr.FromInt(intTargetPort)
+
+	// convert from string to the protocol enum
+	switch strProtocol {
+	case "TCP":
+		protocol = v1.ProtocolTCP
+	case "UDP":
+		protocol = v1.ProtocolUDP
+	default:
+		return fmt.Errorf("unknown protocol: %s", strProtocol)
+	}
+
+	// convert from string to the service type enum
+	switch strServiceType {
+	case "ClusterIP":
+		serviceType = v1.ServiceTypeClusterIP
+	case "NodePort":
+		serviceType = v1.ServiceTypeNodePort
+	case "LoadBalancer":
+		serviceType = v1.ServiceTypeLoadBalancer
+	case "ExternalName":
+		return fmt.Errorf("type: %s not supported", strServiceType)
+	default:
+		return fmt.Errorf("unknown service type: %s", strServiceType)
+	}
+
+	// get the namespace
+	namespace, _, err := o.clientConfig.Namespace()
+	if err != nil {
+		return err
+	}
+
+	// get the client
+	virtClient, err := kubecli.GetKubevirtClientFromClientConfig(o.clientConfig)
+	if err != nil {
+		return fmt.Errorf("cannot obtain KubeVirt client: %v", err)
+	}
+
+	// does a plain quorum read from the apiserver
+	options := k8smetav1.GetOptions{}
+
+	switch vmType {
+	case "vm", "vms", "virtualmachine", "virtualmachines":
+		// get the VM
+		_, err := virtClient.VM(namespace).Get(vmName, options)
+		if err != nil {
+			return fmt.Errorf("error fetching VirtualMachine: %v", err)
+		}
+	case "ovm", "ovms", "offlinevirtualmachine", "offlinevirtualmachines":
+		// get the offline VM
+		_, err := virtClient.OfflineVirtualMachine(namespace).Get(vmName, options)
+		if err != nil {
+			return fmt.Errorf("error fetching OfflineVirtualMachine: %v", err)
+		}
+	case "vmrs", "vmrss", "virtualmachinereplicaset", "virtualmachinereplicasets":
+		// get the VM replica set
+		vmrs, err := virtClient.ReplicaSet(namespace).Get(vmName, options)
+		if err != nil {
+			return fmt.Errorf("error fetching VirtualMachine ReplicaSet: %v", err)
+		}
+		if len(vmrs.Spec.Selector.MatchExpressions) > 0 {
+			return fmt.Errorf("cannot expose VirtualMachine ReplicaSet with match expressions")
+		}
+		vmName = vmrs.Spec.Selector.MatchLabels["kubevirt.io/vmReplicaSet"]
```

Comment:
In this case you have to use `spec.template.Labels` from the replicaset.

Reply:
if I try to follow kubectl. in: ```kubernetes/pkg/kubectl/cmd/util/factory_client_access.go``` in: ```MapBasedSelectorForObject()``` they use: ```Spec.Selector.MatchLabels``` for the replica set

Reply:
yes you are correct. `spec.selector.matchLabels` is the right thing. Thx.